mysql/versions.go

```diff
@@ -0,0 +1,113 @@
+package mysql
+
+import (
+	"database/sql"
+	"github.com/hashicorp/go-version"
+	"strings"
+)
+
+type SQLVendor string
+
+const (
+	MySQL   SQLVendor = "MySQL"
+	MariaDB SQLVendor = "MariaDB"
+)
+
+type ServerVersion struct {
+	vendor  SQLVendor
+	version *version.Version
+}
+
+func serverVersion(db *sql.DB) (ServerVersion, error) {
+	versionString, err := serverVersionString(db)
+	if err != nil {
+		return ServerVersion{}, err
+	}
+
+	return parseServerVersionString(versionString), nil
+}
+
+func (serverVersion ServerVersion) supportsRoles() bool {
+	var requiredVersion *version.Version
+	if serverVersion.vendor == MariaDB {
+		requiredVersion, _ = version.NewVersion("10.0.5")
+	} else {
+		requiredVersion, _ = version.NewVersion("8.0.0")
+	}
+
+	return serverVersion.version.GreaterThanOrEqual(requiredVersion)
+}
+
+// ALTER USER syntax introduced in MySQL 5.7.6 deprecates SET PASSWORD (GH-8230)
+// We assume that MariaDB will eventually follow suit
+func (serverVersion ServerVersion) deprecatedSetPassword() bool {
+	var requiredVersion *version.Version
+	if serverVersion.vendor == MariaDB {
+		requiredVersion, _ = version.NewVersion("10.2.0")
+	} else {
+		requiredVersion, _ = version.NewVersion("5.7.6")
+	}
+
+	return serverVersion.version.GreaterThanOrEqual(requiredVersion)
+}
+
+func (serverVersion ServerVersion) supportsTlsOption() bool {
+	var requiredVersion *version.Version
+	if serverVersion.vendor == MariaDB {
+		// TODO since when exactly has MariaDB had this option?
```

Comment:
I could not find anything about this. If this PR is generally something you want to pursue further, i could find out by trial and error.

Reply:
The tests on MariaDB 10.1 indicated that it is not supported there, so i put `10.2.0` as the minimally supported version.